Allow more batch actions in the add-on store (remove, enable and disable)

CyrilleB79 opened the pull request:
### Link to issue number:
Closes #15623
### Summary of the issue:
In the store, install action can be executed on multiple add-ons. Other batch actions are needed.
### Description of user facing changes
The following additional batch actions are now available in the store:
* Remove selected add-ons
* Update selected add-ons
* Enable selected add-ons
* Disable selected add-ons

### Description of development approach
* Use the same condition as the one required for the corresponding single action to occur.
* For code simplicity there is no distinction between Enable and Enable (override) for the corresponding multiple action. Anyway, when an add-on should override compatibility to be enabled, there is a confirmation dialog box.
* Also for simplicity, there is no multiple replace action:
  * if a multiple selection contains installable add-on with or without updatable or replaceable add-ons, the "Install selected add-ons" batch action will be available
  * If a multiple selection contains updatable add-ons, replaceable add-ons or both types, the "Update selected add-ons" batch action will be available.
* For batch actions requiring a confirmation (e.g. override incompatibility, uninstall), the confirmation dialog box has a "Remember this choice" checkbox so that the user can answer yes or no for all the same questions.
* When many add-ons are selected and a batch action is applied to them, the action actually applies only to the add-ons for which it is appropriate. E.g batch enable will not be applied to already enabled add-ons.

### Testing strategy:
Manual testing of each of the multiple actions. Also re-tested the corresponding single selection actions.

### Known issues with pull request:
Change log:
I have modified the change log item since batch install was available but I have implemented all the other actions. I have also added my GitHub's name. Unfortunately, it gives the impression that I have made all the job, whereas @seanbudd / NV Access has provided the initial work in #15350. Is it worth thinking to a way to show that an item has been contributed both by NV Access and an external contributor? Or by two external contributors? This would need to be clarified.

### Code Review Checklist:

- [x] Documentation:
  - Change log entry
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] UX of all users considered:
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [x] API is compatible with existing add-ons.
- [x] Security precautions taken.

CyrilleB79 committed:
Allow more batch actions in the add-on store (remove, enable and disable)

amirmahdifard commented:
@CyrilleB79 hi, thanks for this: please also make this for updating addons too: emajin there 8 addons in the updatable addons tab, wich needs update: we have to update them one by one, but if you make this action available to updatable addons as whel, we can select all addons we want to update, and then do update to update them at final. thanks.

CyrilleB79 committed:
Add update batch action

CyrilleB79 committed:
Merge branch 'master' into batchActions

CyrilleB79 committed:
Batch install only if one add-on is Available or Incompatible Available

CyrilleB79 committed:
Display available batch actions according to the selected add-ons

CyrilleB79 committed:
Fix "Replace" always present

CyrilleB79 committed:
Fix issue when batch disabling add-ons

CyrilleB79 committed:
Merge branch 'master' into batchActions

CyrilleB79 committed:
New process to confirm/decline multiple actions for install incompatible, uninstall and enable incompatible: use a "Remember my choice" checkbox.

CyrilleB79 committed:
Some code clean-up

CyrilleB79 committed:
Fix checkPot

CyrilleB79 committed:
Allow batch update in same conditions as single update

CyrilleB79 committed:
Add batch replace action

CyrilleB79 committed:
Documentaztion: Change log and User Guide

CyrilleB79 committed:
Merge branch 'master' into batchActions

seanbudd commented on source/gui/_addonStoreGui/controls/actions.py:
```diff
@@ -140,7 +140,94 @@ def _actions(self) -> List[BatchAddonActionVM]:
+	def __init__(self, addonsList: List[AddonListItemVM]):
+		self.addonsList = addonsList
+
+	def canUseInstallAction(self):
```
can `bool` return types be added to all these functions

seanbudd commented on source/gui/_addonStoreGui/controls/messageDialogs.py:
```diff
@@ -32,6 +32,10 @@
 
 
 class ErrorAddonInstallDialogWithYesNoButtons(ErrorAddonInstallDialog):
+	def __init__(self, *args, useRememberChoiceCheckbox=False, **kwargs):
```
```suggestion
	def __init__(self, *args, useRememberChoiceCheckbox: bool = False, **kwargs):
```

seanbudd commented on source/gui/_addonStoreGui/controls/messageDialogs.py:
```diff
@@ -58,11 +62,27 @@ def _addButtons(self, buttonHelper: ButtonHelper) -> None:
+			)
+			contentsSizer.addItem(self.rememberChoiceCheckbox)
+
+	def shouldRememberChoice(self):
```
```suggestion
	def shouldRememberChoice(self) -> bool:
```

seanbudd commented on source/gui/_addonStoreGui/viewModels/addonList.py:
```diff
@@ -124,6 +124,42 @@ def status(self, newStatus: AvailableAddonStatus):
 	def Id(self) -> str:
 		return self._model.listItemVMId
 
+	def canUseInstallAction(self):
```
can these get bool return types added

seanbudd commented on source/gui/_addonStoreGui/viewModels/store.py:
```diff
@@ -264,22 +245,75 @@ def helpAddon(self, listItemVM: AddonListItemVM) -> None:
 		assert listItemVM.model
-		if _shouldProceedToRemoveAddonDialog(listItemVM.model):
+		if askConfirmation:
+			res, shouldRememberChoice = _shouldProceedToRemoveAddonDialog(
```
```suggestion
			shouldRemove, shouldRememberChoice = _shouldProceedToRemoveAddonDialog(
```

seanbudd commented on source/gui/_addonStoreGui/viewModels/store.py:
```diff
@@ -264,22 +245,75 @@ def helpAddon(self, listItemVM: AddonListItemVM) -> None:
+		return res, shouldRememberChoice
+
+	def removeAddons(self, listItemVMs: Iterable[AddonListItemVM[_AddonStoreModel]]) -> None:
+		shoulRemove = True
```
```suggestion
		shouldRemove = True
```

seanbudd replied:
this is still unresolved

seanbudd commented on source/gui/_addonStoreGui/viewModels/store.py:
```diff
@@ -264,22 +245,75 @@ def helpAddon(self, listItemVM: AddonListItemVM) -> None:
 		from gui import mainFrame
-		if _shouldInstallWhenAddonTooOldDialog(mainFrame, listItemVM.model):
+		if askConfirmation:
+			res, shouldRememberChoice = _shouldInstallWhenAddonTooOldDialog(
```
```suggestion
			shouldInstall, shouldRememberChoice = _shouldInstallWhenAddonTooOldDialog(
```

seanbudd commented on source/gui/_addonStoreGui/viewModels/store.py:
```diff
@@ -264,22 +245,75 @@ def helpAddon(self, listItemVM: AddonListItemVM) -> None:
 			addonDataManager._deleteCacheInstalledAddon(listItemVM.model.name)
 			assert listItemVM.model._addonHandlerModel is not None
 			listItemVM.model._addonHandlerModel.requestRemove()
 			self.refresh()
```
should refreshed be moved here too like in the changes in `installOverrideIncompatibilityForAddon`

seanbudd replied:
This is unanswered

```suggestion
		self.refresh()
```

CyrilleB79 replied:
> should refreshed be moved here too like in the changes in `installOverrideIncompatibilityForAddon`

No, that's the other way around, i.e. `installOverrideIncompatibilityForAddon` to be fixed. There is no need to refresh if no action has occurred.

seanbudd commented on source/gui/_addonStoreGui/viewModels/store.py:
```diff
@@ -313,22 +347,122 @@ def _handleEnableDisable(self, listItemVM: AddonListItemVM[_AddonManifestModel],
 		from ... import mainFrame
-		if _shouldEnableWhenAddonTooOldDialog(mainFrame, listItemVM.model):
+		if askConfirmation:
+			res, shouldRememberChoice = _shouldEnableWhenAddonTooOldDialog(
```
```suggestion
			shouldEnable, shouldRememberChoice = _shouldEnableWhenAddonTooOldDialog(
```

seanbudd commented on source/gui/_addonStoreGui/viewModels/store.py:
```diff
@@ -313,22 +347,122 @@ def _handleEnableDisable(self, listItemVM: AddonListItemVM[_AddonManifestModel],
+		shouldEnableIncompatible = True
+		shouldRememberChoice = False
+		for aVM in listItemVMs:
+			if aVM.status in (
+				AvailableAddonStatus.INCOMPATIBLE_DISABLED,
+				AvailableAddonStatus.PENDING_INCOMPATIBLE_DISABLED,
+			) and not aVM.model.isCompatible and aVM.model.canOverrideCompatibility:
```
```suggestion
			if aVM.canUseEnableOverrideIncompatibilityAction() and not aVM.model.isCompatible:
```

seanbudd replied:
the change made by you in response to this review comment dropped the usage of `and not aVM.model.isCompatible`, is this intentional?

seanbudd commented on source/gui/_addonStoreGui/viewModels/store.py:
```diff
@@ -313,22 +347,122 @@ def _handleEnableDisable(self, listItemVM: AddonListItemVM[_AddonManifestModel],
+						askConfirmation=True,
+						useRememberChoiceCheckbox=True,
+					)
+			elif aVM.status in (
+				AvailableAddonStatus.DISABLED,
+				AvailableAddonStatus.PENDING_DISABLE,
+			):
```
```suggestion
			elif aVM.canUseEnableAction():
```

seanbudd commented on source/gui/_addonStoreGui/viewModels/store.py:
```diff
@@ -313,22 +347,122 @@ def _handleEnableDisable(self, listItemVM: AddonListItemVM[_AddonManifestModel],
-	def replaceAddon(self, listItemVM: AddonListItemVM) -> None:
+	def disableAddons(self, listItemVMs: Iterable[AddonListItemVM[_AddonStoreModel]]) -> None:
+		for aVM in listItemVMs:
+			if aVM.status not in [
```
this doesn't quite match `canUseDisableAction`. Should we just use that here?

CyrilleB79 replied:
Good catch!

seanbudd commented on source/gui/_addonStoreGui/viewModels/store.py:
```diff
@@ -313,22 +347,122 @@ def _handleEnableDisable(self, listItemVM: AddonListItemVM[_AddonManifestModel],
-		if _shouldProceedWhenInstalledAddonVersionUnknown(mainFrame, listItemVM.model):
+		assert listItemVM.model
+		if askConfirmation:
+			res, shouldRememberChoice = _shouldProceedWhenInstalledAddonVersionUnknown(
```
```suggestion
			shouldReplace, shouldRememberChoice = _shouldProceedWhenInstalledAddonVersionUnknown(
```

seanbudd commented on source/gui/_addonStoreGui/viewModels/store.py:
```diff
@@ -313,22 +347,122 @@ def _handleEnableDisable(self, listItemVM: AddonListItemVM[_AddonManifestModel],
+						self.replaceAddon(aVM, askConfirmation=False)
+					else:
+						log.debug(
+							f"Skipping {aVM.Id} as replacement has been previously declined for all remaining  add-ons."
```
```suggestion
							f"Skipping {aVM.Id} as replacement has been previously declined for all remaining add-ons."
```

seanbudd commented on user_docs/en/changes.t2t:
```diff
@@ -8,7 +8,7 @@ What's New in NVDA
 == New Features ==
 - Added support for Bluetooth Low Energy HID Braille displays. (#15470)
-- The Add-on Store now supports installing add-ons in bulk by selecting multiple add-ons. (#15350)
+- The Add-on Store now supports actions (installing, enabling add-ons, etc.) in bulk by selecting multiple add-ons. (#15350, #15623, @CyrilleB79)
```
```suggestion
- The Add-on Store now supports bulk actions (e.g. installing, enabling add-ons) by selecting multiple add-ons. (#15350, #15623, @CyrilleB79)
```

seanbudd replied:
>I have modified the change log item since batch install was available but I have implemented all the other actions. I have also added my GitHub's name. Unfortunately, it gives the impression that I have made all the job, whereas @seanbudd / NV Access has provided the initial work in https://github.com/nvaccess/nvda/pull/15350. Is it worth thinking to a way to show that an item has been contributed both by NV Access and an external contributor? Or by two external contributors? This would need to be clarified.

I wouldn't worry about this, we don't need to be credited.

CyrilleB79 replied:
In the current PR, that's not so much a problem since I understand the initial PR (#15350). I was just mentioning since in an hypothetic future case, I (or another external contributor) may be associated this way with a PR that I do not fully understand or for which I have not the background.

seanbudd reviewed (commented):
Changes generally look good to me

CyrilleB79 committed:
Address review comments and more

CyrilleB79 committed:
Rewrite log messages

CyrilleB79 commented:
@seanbudd, I have applied review comments. Moreover, I have use more `canUseXxxAction` when possible and rephrased the log messages.

I am still concerned with one point:
In your previous work, batch install was operating either on available add-ons and on updatable add-ons. In the initial description of this PR, I indicate that I handle batch install, update and replace separately for simplicity but that's not true for install since I have kept your original condition (available or updatable).
After thinking again to this, maybe the best solution would be to commonalize these three task and just provide:
* "Install selected add-ons" in "Available add-ons tab to operate on available, updatable and migratable add-ons, i.e. `canUseInstallAction` or `canUseUpdateAction` or `canUseReplaceAction`
* "Update selected add-ons" in other tabs

The limitation of this solution are:
1. You can have "Install selected add-ons" on a selection that contains only updatable or migratable add-ons. I think that it is understandable anyway.
2. You can have "Update selected add-ons" on a selection containing only migratable add-ons. I think that's not a problem since there is the dialog box recording that the versions cannot be compared.

Do you agree with this approach? If no, what do you suggest instead?

seanbudd commented:
The approach you describe sounds good, I don't think multiple actions are required for install vs. update vs. replace

CyrilleB79 committed:
Addressing review

CyrilleB79 committed:
Rework multiple action availability

CyrilleB79 committed:
Merge branch 'master' into batchActions

CyrilleB79 commented:
@seanbudd, at the end, I have implemented a mid-way solution regarding the install/update/migrate batch actions. It allows to have update batch action when only updatable or migratable (replaceable) add-ons are selected, but no installable one. I have updated the initial description accordingly; see it for more details.

CyrilleB79 committed:
Remove debug code

lukaszgo1 commented on source/gui/addonStoreGui/viewModels/store.py:
```diff
@@ -313,22 +347,110 @@ def _handleEnableDisable(self, listItemVM: AddonListItemVM[_AddonManifestModel],
+			elif aVM.canUseEnableAction():
+				self.enableAddon(aVM)
+			else:
+				log.debug(
```
This outer call to `log.debug` seems unnecessary.

lukaszgo1 commented on source/gui/addonStoreGui/viewModels/store.py:
```diff
@@ -313,22 +347,110 @@ def _handleEnableDisable(self, listItemVM: AddonListItemVM[_AddonManifestModel],
+					log.debug(f"Skipping {aVM.Id} ({aVM.status}) as it is not relevant for enable action")
+				)
+
+	
```
Unnecessary two blank lines here. I have no idea why this has not been caught by Flake8 though.

CyrilleB79 committed:
Small clean-up

seanbudd reviewed (approved):
Thanks @CyrilleB79

Qchristensen reviewed (approved):
Reads well, good work.
